common/persistence/cassandra/metadata_store.go

```diff
@@ -307,46 +307,65 @@ func (m *MetadataStore) GetNamespace(request *p.GetNamespaceRequest) (*p.Interna
 	}, nil
 }
 
-func (m *MetadataStore) ListNamespaces(request *p.ListNamespacesRequest) (*p.InternalListNamespacesResponse, error) {
+func (m *MetadataStore) ListNamespaces(request *p.InternalListNamespacesRequest) (*p.InternalListNamespacesResponse, error) {
 	query := m.session.Query(templateListNamespaceQueryV2, constNamespacePartition)
-	iter := query.PageSize(request.PageSize).PageState(request.NextPageToken).Iter()
-
+	pageSize := request.PageSize
+	nextPageToken := request.NextPageToken
 	response := &p.InternalListNamespacesResponse{}
+
 	for {
-		var name string
-		var detail []byte
-		var detailEncoding string
-		var notificationVersion int64
-		var isGlobal bool
-		if !iter.Scan(
-			nil,
-			&name,
-			&detail,
-			&detailEncoding,
-			&notificationVersion,
-			&isGlobal,
-		) {
-			// done iterating over all namespaces in this page
-			break
-		}
+		iter := query.PageSize(pageSize).PageState(nextPageToken).Iter()
+		skippedRows := 0
+
+		for {
+			var name string
+			var detail []byte
+			var detailEncoding string
+			var notificationVersion int64
+			var isGlobal bool
+			if !iter.Scan(
+				nil,
+				&name,
+				&detail,
+				&detailEncoding,
+				&notificationVersion,
+				&isGlobal,
+			) {
+				// done iterating over all namespaces in this page
+				break
+			}
 
-		// do not include the metadata record
-		if name != namespaceMetadataRecordName {
+			// do not include the metadata record
+			if name == namespaceMetadataRecordName {
```

Comment:
This file changes are not related to deleted namespace but it had the same problem (essentially bug). If there was `namespaceMetadataRecordName` record in response, results might have fewer items than defined by `pageSize`. It is easy testable if `pageSize` set to `1`. Previous implementation just returned empty page in the middle of pagination.